### ISSUE TITLE
how to interpret the final best architecture after the neural architecture search

### ISSUE THREAD

**Issue by khatiakilanava** [microsoft/nni#915]
**Short summary about the issue/question**:
I am performing the best neural architecture search based on the example in network_morphism folder, cifar10_Keras. Is there any way to interpret the best architecture found so that I can code a similar architecture using keras, in python later?
**Brief what process you are following**: 
I'm performing a binary classification. I have training/testing/validation folders separately with images 32x32x3, I have modified cifar10_keras example a little bit: I have specified my own input folders, I am evaluating the trained model on validation and testing folders(net.evaluate originally was performed on the validation set only) and I have set the number of output nodes as 2 in config.yml
I am having the difficulty of interpreting the final results I get. As I understand, architectures are saved in experiments/myexperiment_id/log/model_path. In that path: the output of best_model.txt  is 0. Does it mean that the best architecture is saved in 0.json? And if this is correct, the parameters in 0.json look like this:
{"input_shape": [32, 32, 3], "vis": null, "weighted": false, "operation_history": [], "layer_id_to_input_node_ids": {"0": [0], "1": [1], "2": [2], "3": [3], "4": [4], "5": [5], "6": [6], "7": [7], "8": [8], "9": [9], "10": [10], "11": [11], "12": [12], "13": [13], "14": [14], "15": [15], "16": [16]}, "layer_id_to_output_node_ids": {"0": [1], "1": [2], "2": [3], "3": [4], "4": [5], "5": [6], "6": [7], "7": [8], "8": [9], "9": [10], "10": [11], "11": [12], "12": [13], "13": [14], "14": [15], "15": [16], "16": [17]}, "adj_list": {"0": [[1, 0]], "1": [[2, 1]], "2": [[3, 2]], "3": [[4, 3]], "4": [[5, 4]], "5": [[6, 5]], "6": [[7, 6]], "7": [[8, 7]], "8": [[9, 8]], "9": [[10, 9]], "10": [[11, 10]], "11": [[12, 11]], "12": [[13, 12]], "13": [[14, 13]], "14": [[15, 14]], "15": [[16, 15]], "16": [[17, 16]], "17": []}, "reverse_adj_list": {"0": [], "1": [[0, 0]], "2": [[1, 1]], "3": [[2, 2]], "4": [[3, 3]], "5": [[4, 4]], "6": [[5, 5]], "7": [[6, 6]], "8": [[7, 7]], "9": [[8, 8]], "10": [[9, 9]], "11": [[10, 10]], "12": [[11, 11]], "13": [[12, 12]], "14": [[13, 13]], "15": [[14, 14]], "16": [[15, 15]], "17": [[16, 16]]}, "node_list": [[0, [32, 32, 3]], [1, [32, 32, 3]], [2, [32, 32, 3]], [3, [32, 32, 64]], [4, [16, 16, 64]], [5, [16, 16, 64]], [6, [16, 16, 64]], [7, [16, 16, 64]], [8, [8, 8, 64]], [9, [8, 8, 64]], [10, [8, 8, 64]], [11, [8, 8, 64]], [12, [4, 4, 64]], [13, [64]], [14, [64]], [15, [64]], [16, [64]], [17, [2]]], "layer_list": [[0, ["StubReLU", 0, 1]], [1, ["StubBatchNormalization2d", 1, 2, 3]], [2, ["StubConv2d", 2, 3, 3, 64, 3, 1, 1]], [3, ["StubPooling2d", 3, 4, 2, 2, 0]], [4, ["StubReLU", 4, 5]], [5, ["StubBatchNormalization2d", 5, 6, 64]], [6, ["StubConv2d", 6, 7, 64, 64, 3, 1, 1]], [7, ["StubPooling2d", 7, 8, 2, 2, 0]], [8, ["StubReLU", 8, 9]], [9, ["StubBatchNormalization2d", 9, 10, 64]], [10, ["StubConv2d", 10, 11, 64, 64, 3, 1, 1]], [11, ["StubPooling2d", 11, 12, 2, 2, 0]], [12, ["StubGlobalPooling2d", 12, 13]], [13, ["StubDropout2d", 13, 14, 0.25]], [14, ["StubDense", 14, 15, 64, 64]], [15, ["StubReLU", 15, 16]], [16, ["StubDense", 16, 17, 64, 2]]]}

How can I interpret this? I was willing to find the best architecture and then implement the same architecture in python using keras.  Is there any way to do it easily? 

I was willing to find the best architecture, code it in python, keras and then tune it using again NNI.

**How to reproduce it**: 

**nni Environment**:
- nni version:0.5.2
- nni mode(local|pai|remote):local
- OS: Ubuntu 16
- python version: 3.6
- is conda or virtualenv used?: no
- is running in docker?: no

**Anything else we need to know**:

**Comment by leelaylay** [comment]
@khatiakilanava 
## Find best model ID
The best model ID can be found in `best_model.txt`. Also, you can find ID of the model with best metric value in WebUI (usually in 8080 port).

## Tune the architecture found by Network Morphism
There are several ways that can reuse the architecture again as far as I know.
1. Use NNI API
```python
import nni
from nni.networkmorphism_tuner.graph import json_to_graph

def build_graph_from_json(ir_model_json):
    """build model from json representation
    """
    graph = json_to_graph(ir_model_json)
    logging.debug(graph.operation_history)
    model = graph.produce_keras_model()
    return model

json_of_model = "the_json_string_of_model"
keras_model = build_graph_from_json(json_of_model)
# than you can tune the keras model with your own code
```
2. Use keras API

You can save model and its weight in your trial code as I mentioned in #848 
```python
# it's better to save model with id in nni local mode
model_id = nni.get_sequence_id()
# serialize model to JSON
model_json = model.to_json()
with open("model-{}.json".format(model_id), "w") as json_file:
    json_file.write(model_json)
# serialize weights to HDF5
model.save_weights("model-{}.h5".format(model_id))
print("Saved model to disk")
```
After that you can load the model by keras API.
```python
# load json and create model
with open('model-ID.json', 'r') as json_file:
    loaded_model_json = json_file.read()
loaded_model = model_from_json(loaded_model_json)
# load weights into new model
loaded_model.load_weights("model.h5")
print("Loaded model from disk")
```
If you want to get a view of the complex model, you can also use utility function from Keras API.
```python
from keras.utils import plot_model
plot_model(model, to_file='model.png')
```

**Comment by xuehui1991** [comment]
@khatiakilanava You may check this reference : )
https://github.com/Microsoft/nni/blob/master/docs/en_US/networkmorphismTuner.md

**Comment by khatiakilanava** [comment]
Thank you all for your answers. @leelaylay I can run the experiment successfully across all trials but I can't load webui, while I paste the link into the browser(http://172.17.64.219:8080/), it never loads and it says: The server at 172.17.64.219  is taking too long to respond. Do you know what might be a problem?

**Comment by lvybriage** [comment]
@khatiakilanava 
Thanks for your response. I'm sorry to reply late.
It looks webui is error. First, I want to know restAPI's status. You can use the command "nnictl trial ls" to check whether rest is blocked. If you can see the command return trial message, I'll check the webui.

**Comment by khatiakilanava** [comment]
@lvybriage 
Thanks for your reply. It looks like this:
I don't see the problem :/
[
    {
        "endTime":"2019/03/29 04:32:46",
        "finalMetricData":[
            {
                "data":"0.9453560468204902",
                "parameterId":"0",
                "sequence":0,
                "timestamp":1553833961570,
                "trialJobId":"yHnaI",
                "type":"FINAL"
            }
        ],
        "hyperParameters":[
            "{\"parameter_id\":0,\"parameter_source\":\"algorithm\",\"parameters                                                                             \":{\"batch_size\":32,\"optimizer\":\"SGD\",\"learning_rate\":0.0001}}"
        ],
        "id":"yHnaI",
        "logPath":"file://localhost:/home/ubuntu/nni/experiments/ZFJJLzfC/trials                                                                             /yHnaI",
        "sequenceId":0,
        "startTime":"2019/03/28 22:18:19",
        "status":"SUCCEEDED"
    },
    {
        "endTime":"2019/03/29 05:12:34",
        "finalMetricData":[
            {
                "data":"0.944447923846659",
                "parameterId":"1",
                "sequence":0,
                "timestamp":1553836351654,
                "trialJobId":"ASICa",
                "type":"FINAL"
            }
        ],

**Comment by lvybriage** [comment]
@khatiakilanava So, what browser do you use? And please check whether the link "http://172.17.64.219:8080/api/v1/nni/trial-jobs"(your IP/api/v1/nni/trial-jobs) in the browser's address can get message.

**Comment by scarlett2018** [comment]
Closing no-responds issues within 30 days. @khatiakilanava - please reopen the issue if you are still seeing the problem on your side. (it is highly recommended to install a latest version as well).
